lib/storage/data/file/store.js

```diff
@@ -0,0 +1,274 @@
+                return next();
+            });
+        },
+        err => {
+            assert.strictEqual(err, null, `Error creating data files ${err}`);
+            this.logger.info('data file store init complete, ' +
+                             'go forth and store data.');
+            callback();
+        });
+    }
+
+    /**
+     * Get the filesystem path to a stored object file from its key
+     *
+     * @param {String} key - the object key
+     * @return {String} the absolute path to the file containing the
+     *   object contents
+     */
+    getFilePath(key) {
+        const hash = stringHash(key);
+        const folderHashPath = ((hash % FOLDER_HASH)).toString();
+        return `${this.dataPath}/${folderHashPath}/${key}`;
+    }
+
+    /**
+     * Put a new object to the storage backend
+     *
+     * @param {stream.Readable} dataStream - input stream providing the
+     *   object data
+     * @param {Number} size - Total byte size of the data to put
+     * @param {Logger} log - logging object
+     * @param {DataFileStore~putCallback} callback - called when done
+     * @return {undefined}
+     */
+    put(dataStream, size, log, callback) {
+        const key = crypto.pseudoRandomBytes(20).toString('hex');
+        const filePath = this.getFilePath(key);
+        log.debug('starting to write data', { key, filePath });
+        dataStream.pause();
+        fs.open(filePath, 'wx', (err, fd) => {
+            if (err) {
+                log.error('error opening filePath', { error: err });
+                return callback(err);
+            }
+            const cbOnce = jsutil.once(callback);
+            // disable autoClose so that we can close(fd) only after
+            // fsync() has been called
+            const fileStream = fs.createWriteStream(filePath,
+                                                    { fd,
+                                                      autoClose: false });
+
+            fileStream.on('finish', () => {
+                function ok() {
+                    log.debug('finished writing data', { key });
+                    return cbOnce(null, key);
+                }
+                if (this.noSync) {
+                    fs.close(fd);
+                    return ok();
+                }
+                fs.fsync(fd, err => {
+                    fs.close(fd);
+                    if (err) {
+                        log.error('fsync error', { error: err });
+                        return cbOnce(err);
+                    }
+                    return ok();
+                });
+                return undefined;
+            }).on('error', err => {
+                log.error('error streaming data on write',
+                          { error: err });
+                // destroying the write stream forces a close(fd)
+                fileStream.destroy();
+                return cbOnce(err);
+            });
+            dataStream.resume();
+            dataStream.pipe(fileStream);
```

Comment:
Instead of making `pipe` end `fileStream` automatically, why not call `fd.close()` explicitly after `pipe`. This also ensures that the callback is not called twice.